src/Cake.Core/ICakeEngine.cs

```diff
@@ -20,25 +20,73 @@ public interface ICakeEngine
-        /// Raised during teardown after all other tasks have been run.
+        /// Raised at the start of setup before any tasks are run.
         /// </summary>
+        event EventHandler<BeforeSetupEventArgs> BeforeSetup;
+
+        /// <summary>
+        /// Raised at the end of setup before any tasks are run.
+        /// </summary>
+        event EventHandler<AfterSetupEventArgs> AfterSetup;
+
+        /// <summary>
+        /// [deprecated] Teardown is obsolete and will be removed in a future release. Use <see cref="BeforeTeardown" /> instead.
+        /// Raised at the start of teardown after all other tasks have been run.
+        /// </summary>
+        [Obsolete("Teardown is obsolete and will be removed in a future release. Use BeforeTeardown instead.")]
         event EventHandler<TeardownEventArgs> Teardown;
 
         /// <summary>
-        /// Raised before each task is run.
+        /// Raised at the start of teardown after all other tasks have been run.
+        /// </summary>
+        event EventHandler<BeforeTeardownEventArgs> BeforeTeardown;
+
+        /// <summary>
+        /// Raised at the end of teardown after all other tasks have been run.
+        /// </summary>
+        event EventHandler<AfterTeardownEventArgs> AfterTeardown;
+
+        /// <summary>
+        /// [deprecated] TaskSetup is obsolete and will be removed in a future release. Use <see cref="BeforeTaskSetup" /> instead.
+        /// Raised at the start of task setup before each task is run.
         /// </summary>
+        [Obsolete("TaskSetup is obsolete and will be removed in a future release. Use BeforeTaskSetup instead.")]
         event EventHandler<TaskSetupEventArgs> TaskSetup;
 
         /// <summary>
-        /// Raised after each task has been run.
+        /// Raised at the start of task setup before each task is run.
         /// </summary>
+        event EventHandler<BeforeTaskSetupEventArgs> BeforeTaskSetup;
+
+        /// <summary>
+        /// Raised at the end of task setup before each task is run.
+        /// </summary>
+        event EventHandler<AfterTaskSetupEventArgs> AfterTaskSetup;
+
+        /// <summary>
+        /// [deprecated] TaskTeardown is obsolete and will be removed in a future release. Use <see cref="BeforeTaskTeardown" /> instead.
+        /// Raised at the start of task teardown after each task has been run.
+        /// </summary>
+        [Obsolete("TaskTeardown is obsolete and will be removed in a future release. Use BeforeTaskTeardown instead.")]
         event EventHandler<TaskTeardownEventArgs> TaskTeardown;
 
+        /// <summary>
+        /// Raised at the start of task teardown after each task has been run.
+        /// </summary>
+        event EventHandler<BeforeTaskTeardownEventArgs> BeforeTaskTeardown;
+
+        /// <summary>
+        /// Raised at the end of task teardown after each task has been run.
+        /// </summary>
+        event EventHandler<AfterTaskTeardownEventArgs> AfterTaskTeardown;
```

Comment:
Do we want to expose these events in the interface `ICakeEngine`?

Reply:
These need to be on the interface to allow for things like the teamcity logger to log "end blocks" as service messages.